File: skills/lasersell/examples/build-and-send-sell/build_and_send_sell.ts
/**
 * Build, sign, and submit a sell transaction.
 *
 * Flow:
 * 1. Build unsigned tx with LaserSell API.
 * 2. Sign tx using a local Solana keypair JSON file.
 * 3. Send tx via either standard RPC or Helius Sender.
 *
 * Before running:
 * - Replace all placeholders, including `keypairPath`.
 * - For `sendTarget = "rpc"`, provide a valid `rpcUrl`.
 */
import { readFile } from "node:fs/promises";

import { Keypair } from "@solana/web3.js";
import {
  ExitApiClient,
  sendTransaction,
  sendTargetHeliusSender,
  sendTargetRpc,
  signUnsignedTx,
  type BuildSellTxRequest,
} from "@lasersell/lasersell-sdk";

async function main(): Promise<void> {
  const apiKey = "REPLACE_WITH_API_KEY";
  const keypairPath = "REPLACE_WITH_KEYPAIR_PATH";
  const sendTargetInput: string = "rpc";
  const sendTarget: "rpc" | "helius_sender" =
    sendTargetInput === "helius_sender" ? "helius_sender" : "rpc";
  const rpcUrl = "REPLACE_WITH_RPC_URL";

  const keypair = await readKeypairFile(keypairPath);

  const client = ExitApiClient.withApiKey(apiKey);
  const request: BuildSellTxRequest = {
    mint: "REPLACE_WITH_MINT",
    user_pubkey: "REPLACE_WITH_WALLET_PUBKEY",
    amount_tokens: 1_000_000,
    slippage_bps: 2_000,
    output: "SOL",
  };

  const unsignedTxB64 = await client.buildSellTxB64(request);
  const signedTx = signUnsignedTx(unsignedTxB64, keypair);

  const target =
    sendTarget === "helius_sender"
      ? sendTargetHeliusSender()
      : sendTargetRpc(rpcUrl);
  const signature = await sendTransaction(target, signedTx);

  console.log(`signature=${signature}`);
}

async function readKeypairFile(path: string): Promise<Keypair> {
  const raw = await readFile(path, "utf8");
  const parsed = JSON.parse(raw) as unknown;
  if (!Array.isArray(parsed)) {
    throw new Error("keypair file must be an array of numbers");
  }

  const bytes = Uint8Array.from(parsed.map((value) => Number(value)));
  return Keypair.fromSecretKey(bytes);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: skills/lasersell/examples/build-and-send-sell/build_and_send_sell.py
"""Build, sign, and submit a sell transaction.

Flow:
1. Build unsigned sell tx via LaserSell API.
2. Sign tx using local keypair.
3. Submit tx via standard RPC or Helius Sender.

Before running:
- Replace all placeholders, including keypair path and API key.
- Set `send_target` to `"rpc"` or `"helius_sender"`.
- If using `"rpc"`, set `rpc_url` to a valid Solana RPC endpoint.
"""

from __future__ import annotations

import asyncio
import json
from pathlib import Path

from solders.keypair import Keypair

from lasersell_sdk.exit_api import BuildSellTxRequest, ExitApiClient, SellOutput
from lasersell_sdk.tx import SendTargetHeliusSender, SendTargetRpc, send_transaction, sign_unsigned_tx


async def main() -> None:
    api_key = "REPLACE_WITH_API_KEY"
    keypair_path = Path("REPLACE_WITH_KEYPAIR_PATH")
    send_target = "rpc"  # "rpc" or "helius_sender"
    rpc_url = "REPLACE_WITH_RPC_URL"

    keypair = read_keypair_file(keypair_path)
    exit_api = ExitApiClient.with_api_key(api_key)

    request = BuildSellTxRequest(
        mint="REPLACE_WITH_MINT",
        user_pubkey="REPLACE_WITH_WALLET_PUBKEY",
        amount_tokens=1_000_000,
        slippage_bps=2_000,
        output=SellOutput.SOL,
    )

    unsigned_tx_b64 = await exit_api.build_sell_tx_b64(request)
    signed_tx = sign_unsigned_tx(unsigned_tx_b64, keypair)

    if send_target == "helius_sender":
        target = SendTargetHeliusSender()
    elif send_target == "rpc":
        target = SendTargetRpc(url=rpc_url)
    else:
        raise ValueError(f"send_target must be rpc or helius_sender (got {send_target!r})")

    signature = await send_transaction(target, signed_tx)

    print(f"signature={signature}")


def read_keypair_file(path: Path) -> Keypair:
    raw = path.read_text(encoding="utf-8")
    parsed = json.loads(raw)
    if not isinstance(parsed, list):
        raise ValueError("keypair file must be an array of numbers")

    secret = bytes(int(value) for value in parsed)
    return Keypair.from_bytes(secret)


if __name__ == "__main__":
    asyncio.run(main())
